### PR TITLE
Skip incomplete format specifier

### PR DESCRIPTION
**Pull request by ganmacs**
Make ruby's `printf` raise an `ArgumentError` when format string is incomplete.

https://bugs.ruby-lang.org/issues/13315

In C language, `printf("%")` shows warning as follow.

```c
#include <stdio.h>

int main(void)
{
  printf("%");
  return 0;
}
```

```
5:11: warning: incomplete format specifier [-Wformat]
  printf("%");
          ^
1 warning generated.
```

I think it's nice to notify that incomplete format is invalid for developer in ruby too.
Since this change will drop displayed `"%"` silently, Raising an error is better than showing a warning.

### CODE REVIEW

File: sprintf.c [hunk 1]
```diff
@@ -519,8 +519,11 @@ rb_str_format(int argc, const VALUE *argv, VALUE fmt)
 	VALUE sym = Qnil;
 
 	for (t = p; t < end && *t != '%'; t++) ;
-	if (t + 1 == end) ++t;
 	PUSH(p, t - p);
+	if (t + 1 == end) {
+	    if (*t == '%') rb_warn("incomplete format specifier");
+	     ++t;
```
Comment (sorah) [inline sprintf.c]: indent

File: sprintf.c [hunk 2]
```diff
@@ -519,8 +519,11 @@ rb_str_format(int argc, const VALUE *argv, VALUE fmt)
 	VALUE sym = Qnil;
 
 	for (t = p; t < end && *t != '%'; t++) ;
-	if (t + 1 == end) ++t;
 	PUSH(p, t - p);
+	if (t + 1 == end) {
+	    if (*t == '%') rb_warn("incomplete format specifier");
```
Comment (sorah) [inline sprintf.c]: How about using `rb_warning` (it shows warning only in verbose mode)

File: test/ruby/test_sprintf.rb [hunk 3]
```diff
@@ -372,6 +372,11 @@ def test_escape
     assert_equal("%" * BSIZ, sprintf("%%" * BSIZ))
   end
 
+  def test_percent_sing_at_end
```
Comment (mrkn) [inline test/ruby/test_sprintf.rb]: It may be a typo: s/sing/sign/

Comment (ganmacs) [inline test/ruby/test_sprintf.rb]: Sorry, fixed.
https://github.com/ruby/ruby/pull/1560/commits/1b57250073c77f34c3503fa86a7d02aa170b7764

**Comment by sorah** [comment]
How about making this as an error?

**Comment by nobu** [comment]
ruby/spec also needs update.

File: sprintf.c [hunk 4]
```diff
@@ -519,8 +519,11 @@ rb_str_format(int argc, const VALUE *argv, VALUE fmt)
 	VALUE sym = Qnil;
 
 	for (t = p; t < end && *t != '%'; t++) ;
-	if (t + 1 == end) ++t;
 	PUSH(p, t - p);
```
Comment (eregon) [inline sprintf.c]: Why is the push done before the `++t` increment now?

Comment (nobu) [inline sprintf.c]: To exit the loop after a single `'%'` at the end.
Otherwise, `p` skips it and reaches `end`.
This isn't a problem right now, but if `SHARABLE_MIDDLE_SUBSTRING` in string.c is non-zero, `*end` may not be `'\0'`.

Comment (eregon) [inline sprintf.c]: I committed in fd8df3ab3d935b4a201f9cb28598c95e4bb1ef14 with the original order as this seems an unrelated issue, and the extra exception should not affect this logic.
Please fix if you think it is worth changing.
